File: web/lib/openrouter.ts
/** Everything that talks to OpenRouter, from the visitor's browser only.
 *
 *  The rule this file exists to enforce: **no key ever reaches our server.**
 *  The visitor signs in with their own OpenRouter account, the key that comes
 *  back is theirs, it lives in their localStorage, and every request below is
 *  made by their browser directly to openrouter.ai. Nothing here runs on
 *  Vercel, so there is no server-side path a key could leak through.
 *
 *  That is not a stylistic choice. It is what lets this page exist at all:
 *  inference is billed to the visitor, so a free tool costs the project
 *  nothing, and a 2-hour VOD is ~24 sequential calls that would blow any
 *  serverless timeout but is merely a loop in a tab.
 *
 *  Two OpenRouter quirks worth knowing before editing:
 *
 *  1. `HTTP-Referer` is deliberately NOT the standard `Referer` header, which
 *     browsers forbid scripts from setting. OpenRouter picked a non-standard
 *     name precisely so browser apps can attribute themselves, and their CORS
 *     allow-list includes it. This is what creates our app page in their
 *     marketplace, and it counts traffic to us regardless of whose key paid.
 *  2. STT pricing is reported in DIFFERENT UNITS PER PROVIDER for the same
 *     model (Groq bills whisper-large-v3-turbo per hour, DeepInfra per
 *     second). Never compute a cost estimate by multiplying blindly.
 */

const BASE = "https://openrouter.ai/api/v1";
const AUTH_URL = "https://openrouter.ai/auth";

const KEY_STORAGE = "clipskitty-openrouter-key";
const VERIFIER_STORAGE = "clipskitty-openrouter-verifier";

/** Sent on every inference call.
 *
 *  `HTTP-Referer` must be the production origin: OpenRouter excludes
 *  localhost traffic from the marketplace, so a dev run will never create the
 *  app page no matter what else is right.
 *
 *  `video-gen` is the closest category OpenRouter offers. It means
 *  *generation* rather than editing, which is not quite us, but the category
 *  list is fixed and unrecognised values are dropped silently. */
function attribution(): Record<string, string> {
	return {
		"HTTP-Referer": typeof window === "undefined" ? "" : window.location.origin,
		"X-OpenRouter-Title": "Clips Kitty Web",
		"X-OpenRouter-Categories": "video-gen",
	};
}

// ---- the key -------------------------------------------------------------

export function loadKey(): string | null {
	try {
		return window.localStorage.getItem(KEY_STORAGE);
	} catch {
		// Private windows and blocked site data throw on access rather than
		// returning null. Signed-out is the correct reading of that.
		return null;
	}
}

export function saveKey(key: string): void {
	try {
		window.localStorage.setItem(KEY_STORAGE, key);
	} catch {
		// Non-fatal: the key stays in memory for this session instead.
	}
}

export function clearKey(): void {
	try {
		window.localStorage.removeItem(KEY_STORAGE);
	} catch {
		/* nothing to clear */
	}
}

// ---- OAuth PKCE ----------------------------------------------------------

function base64url(bytes: ArrayBuffer): string {
	return btoa(String.fromCharCode(...new Uint8Array(bytes)))
		.replace(/\+/g, "-")
		.replace(/\//g, "_")
		.replace(/=+$/, "");
}

/** Send the visitor to OpenRouter to authorise this app.
 *
 *  No client_id and no app registration: OpenRouter's flow identifies the app
 *  after the fact by `HTTP-Referer`, which means there is no secret of ours
 *  embedded in a page anyone can read. The verifier is held in sessionStorage
 *  (not local) so it dies with the tab — it is single-use and worthless after
 *  the exchange. */
export async function beginSignIn(callbackUrl: string): Promise<void> {
	const raw = new Uint8Array(32);
	crypto.getRandomValues(raw);
	const verifier = base64url(raw.buffer);

	const digest = await crypto.subtle.digest(
		"SHA-256",
		new TextEncoder().encode(verifier),
	);

	window.sessionStorage.setItem(VERIFIER_STORAGE, verifier);

	const url = new URL(AUTH_URL);
	url.searchParams.set("callback_url", callbackUrl);
	url.searchParams.set("code_challenge", base64url(digest));
	url.searchParams.set("code_challenge_method", "S256");
	window.location.href = url.toString();
}

/** Exchange the returned code for the visitor's own API key.
 *
 *  Codes are single-use and expire after 10 minutes, so a stale one landing
 *  here (a refreshed callback page, a bookmarked URL) is normal rather than
 *  exceptional — it must read as "sign in again", not as a crash. */
export async function completeSignIn(code: string): Promise<string> {
	const verifier = window.sessionStorage.getItem(VERIFIER_STORAGE);
	if (!verifier) {
		throw new Error(
			"This sign-in link has expired. Please start the sign-in again.",
		);
	}

	const res = await fetch(`${BASE}/auth/keys`, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({
			code,
			code_verifier: verifier,
			code_challenge_method: "S256",
		}),
	});

	window.sessionStorage.removeItem(VERIFIER_STORAGE);

	if (!res.ok) {
		throw new Error(
			res.status === 403
				? "That sign-in link has already been used or has expired. Please sign in again."
				: `OpenRouter refused the sign-in (HTTP ${res.status}).`,
		);
	}

	const { key } = (await res.json()) as { key?: string };
	if (!key) throw new Error("OpenRouter returned no key.");

	saveKey(key);
	return key;
}

/** Where the visitor manages spend limits and revocation.
 *
 *  This app cannot cap what the key it receives is allowed to spend — the
 *  exchange endpoint takes no limit parameter and limits are purely
 *  account-side. Linking here is the honest substitute. The page only opens
 *  for the signed-in owner of that key. */
export async function keySettingsUrl(key: string): Promise<string> {
	const digest = await crypto.subtle.digest(
		"SHA-256",
		new TextEncoder().encode(key),
	);
	const hex = Array.from(new Uint8Array(digest))
		.map((b) => b.toString(16).padStart(2, "0"))
		.join("");
	return `https://openrouter.ai/keys/${hex}`;
}

// ---- errors --------------------------------------------------------------

/** Turn an OpenRouter failure into something a creator can act on.
 *
 *  The status codes that actually happen here each have a different remedy,
 *  and "Request failed with status 402" tells someone none of them. */
async function explain(res: Response): Promise<Error> {
	let detail = "";
	try {
		const body = await res.json();
		detail = body?.error?.message ?? "";
	} catch {
		/* a non-JSON error body tells us nothing extra */
	}

	const message =
		res.status === 401
			? "OpenRouter rejected your sign-in. Try signing in again."
			: res.status === 402
				? "Your OpenRouter account is out of credits. Add credits and try again."
				: res.status === 429
					? "OpenRouter is rate-limiting you. Free accounts get 50 requests a day until you have bought $10 of credit — after that it is 1000."
					: res.status === 403
						? "The provider refused this content. Try a different model, or use the desktop app, which runs the AI locally with no content filter."
						: `OpenRouter returned HTTP ${res.status}.`;

	return new Error(detail ? `${message} (${detail})` : message);
}

// ---- account -------------------------------------------------------------

export type KeyInfo = {
	/** True until the account has EVER purchased $10 of credit. Gates 50
	 *  requests/day versus 1000, which is the difference between "one short
	 *  VOD" and "a working day". */
	isFreeTier: boolean;
	usage: number;
	limitRemaining: number | null;
};

export async function keyInfo(key: string): Promise<KeyInfo> {
	const res = await fetch(`${BASE}/key`, {
		headers: { Authorization: `Bearer ${key}` },
	});
	if (!res.ok) throw await explain(res);

	const { data } = await res.json();
	return {
		isFreeTier: Boolean(data?.is_free_tier),
		usage: Number(data?.usage ?? 0),
		limitRemaining:
			data?.limit_remaining === null || data?.limit_remaining === undefined
				? null
				: Number(data.limit_remaining),
	};
}

export type Model = {
	id: string;
	name: string;
	contextLength: number;
	/** USD per million tokens, or null where OpenRouter reports the `-1`
	 *  sentinel it uses for routers whose price varies by whatever they pick. */
	promptPerM: number | null;
	completionPerM: number | null;
	/** Used to keep models that cannot honour `response_format` out of the
	 *  picker. Scoring parses the reply as JSON, so a model without it is not
	 *  a worse choice here — it is a broken one. */
	supportsJson: boolean;
	/** Transcription models are a different list from chat models and must
	 *  never appear in the same dropdown. */
	isTranscription: boolean;
	/** Approximate USD per HOUR of audio, for transcription models only.
	 *
	 *  Approximate because OpenRouter reports speech pricing in more than one
	 *  unit and does not say which. Two clusters are clearly present:
	 *
	 *    per SECOND   whisper-large-v3-turbo 3.33e-6  -> ~$0.012/hr
	 *    per MINUTE   openai/whisper-1       0.006    ->  $0.36/hr
	 *                 deepgram/nova-3        0.0043   ->  $0.26/hr
	 *                 google/chirp-3         0.016    ->  $0.96/hr
	 *
	 *  Each of those matches the provider's own published rate, and every
	 *  model in the catalogue falls on one side or the other of 1e-4 — so that
	 *  is the split used below. It is a heuristic, which is why the UI prefixes
	 *  the figure with "~" and why the REAL number reported by OpenRouter in
	 *  `usage.cost` is shown once a run has actually spent something.
	 *
	 *  Getting the unit wrong would misprice by 60x, so do not present this as
	 *  exact. Token-billed models (those with a `completion` price, such as the
	 *  gpt-4o transcribe pair) are understated here, since the output side
	 *  cannot be predicted from audio length. */
	audioPerHour: number | null;
};

/** Where the per-second and per-minute clusters divide. See `audioPerHour`. */
const PER_SECOND_CEILING = 1e-4;

function toModel(m: Record<string, unknown>): Model {
	const pricing = (m.pricing ?? {}) as Record<string, string>;
	const perM = (raw: string | undefined): number | null => {
		const n = Number(raw);
		return !raw || Number.isNaN(n) || n < 0 ? null : n * 1_000_000;
	};
	const params = Array.isArray(m.supported_parameters)
		? (m.supported_parameters as string[])
		: [];
	const outputs = ((m.architecture ?? {}) as Record<string, unknown>)
		.output_modalities;

	return {
		id: String(m.id),
		name: String(m.name ?? m.id),
		contextLength: Number(m.context_length ?? 0),
		promptPerM: perM(pricing.prompt),
		completionPerM: perM(pricing.completion),
		supportsJson:
			params.includes("response_format") ||
			params.includes("structured_outputs"),
		isTranscription:
			Array.isArray(outputs) && outputs.includes("transcription"),
		audioPerHour: audioRate(pricing.prompt),
	};
}

/** Estimated USD per hour of audio from a raw speech price. */
function audioRate(raw: string | undefined): number | null {
	const n = Number(raw);
	if (!raw || Number.isNaN(n) || n <= 0) return null;
	return n < PER_SECOND_CEILING ? n * 3600 : n * 60;
}

/** Chat models this visitor can actually use.
 *
 *  `/models/user` rather than `/models`: it already applies their own privacy
 *  settings and provider preferences, so we never offer a model their account
 *  would refuse to route to. */
export async function listModels(key: string): Promise<Model[]> {
	const res = await fetch(`${BASE}/models/user`, {
		headers: { Authorization: `Bearer ${key}` },
	});
	if (!res.ok) throw await explain(res);

	const { data } = await res.json();
	return ((data ?? []) as Record<string, unknown>[])
		.map(toModel)
		.filter((m) => m.id);
}

/** Transcription models, from the PUBLIC catalogue.
 *
 *  Deliberately not `/models/user`, which is where the transcription picker
 *  went wrong: that endpoint returns the visitor's chat models and no
 *  speech-to-text ones, so filtering it for transcription produced an empty
 *  list and the picker collapsed to a single fixed option — a dropdown with
 *  nothing to drop down.
 *
 *  `/models?output_modalities=transcription` needs no auth and returns the
 *  full set (19 at the time of writing, every one reporting
 *  `output_modalities: ["transcription"]`), so the picker has something to
 *  offer before the account is even known. */
export async function listTranscriptionModels(): Promise<Model[]> {
	const res = await fetch(
		`${BASE}/models?output_modalities=transcription&limit=50`,
	);
	if (!res.ok) throw await explain(res);

	const { data } = await res.json();
	return ((data ?? []) as Record<string, unknown>[])
		.map(toModel)
		.filter((m) => m.id && m.isTranscription);
}

// ---- transcription -------------------------------------------------------

export type Segment = { start: number; end: number; text: string };

/** Transcribe one audio chunk.
 *
 *  `offsetSeconds` shifts every returned timestamp back onto the original
 *  recording's clock. Chunks are cut at ~60s because upstream providers time
 *  out after 60 seconds of PROCESSING — that limit binds long before the
 *  25 MB upload cap does, so size is never the reason to split. Get this
 *  wrong and every clip after the first chunk is cut in the wrong place. */
export async function transcribeChunk(
	key: string,
	audio: Blob,
	model: string,
	offsetSeconds: number,
	/** Called with what OpenRouter says this request actually cost, in USD.
	 *
	 *  The ground truth the `audioPerHour` estimate cannot be: OpenRouter
	 *  reports speech pricing in units it does not name, so the figure in the
	 *  picker is inferred. This one is measured. */
	onCost?: (usd: number) => void,
): Promise<Segment[]> {
	const form = new FormData();
	// The extension is how the endpoint infers the format, so it has to match
	// what `segmentAudio` actually produces — which is copied AAC in an MP4
	// container, not the decoded WAV this used to send.
	form.append("file", audio, "chunk.m4a");
	form.append("model", model);
	form.append("response_format", "verbose_json");

	const res = await fetch(`${BASE}/audio/transcriptions`, {
		method: "POST",
		headers: { Authorization: `Bearer ${key}`, ...attribution() },
		body: form,
	});
	if (!res.ok) throw await explain(res);

	const body = await res.json();

	const cost = Number(body?.usage?.cost);
	if (Number.isFinite(cost) && cost > 0) onCost?.(cost);

	// `verbose_json` is only honoured by OpenAI-compatible providers; anything
	// else returns plain `{text}`. One un-timed segment covering the chunk is
	// a usable fallback — scoring still works, the boundaries are just coarse.
	if (!Array.isArray(body?.segments)) {
		const text = String(body?.text ?? "").trim();
		return text
			? [{ start: offsetSeconds, end: offsetSeconds + 60, text }]
			: [];
	}

	return body.segments
		.map((s: Record<string, unknown>) => ({
			start: Number(s.start ?? 0) + offsetSeconds,
			end: Number(s.end ?? 0) + offsetSeconds,
			text: String(s.text ?? "").trim(),
		}))
		.filter((s: Segment) => s.text);
}

// ---- generation ----------------------------------------------------------

/** One completion.
 *
 *  Mirrors `LLMBackend.generate()` in the desktop app deliberately: same
 *  single-string-in, single-string-out shape, so the scoring code ported from
 *  `analysis/highlights.py` needs no reshaping and the two stay comparable.
 *
 *  `provider` is the part worth not deleting. `data_collection: "deny"` and
 *  `zdr: true` keep someone's stream transcript away from providers that
 *  retain or train on it — a guarantee made in code rather than left to each
 *  visitor's account settings. `require_parameters` stops OpenRouter routing
 *  to a provider that would quietly ignore `response_format` and hand back
 *  prose where the caller is about to run JSON.parse. */
export async function generate(
	key: string,
	model: string,
	prompt: string,
	{ jsonMode = false }: { jsonMode?: boolean } = {},
): Promise<string> {
	const body: Record<string, unknown> = {
		model,
		messages: [{ role: "user", content: prompt }],
		provider: {
			data_collection: "deny",
			zdr: true,
			require_parameters: true,
		},
	};
	if (jsonMode) body.response_format = { type: "json_object" };

	const res = await fetch(`${BASE}/chat/completions`, {
		method: "POST",
		headers: {
			Authorization: `Bearer ${key}`,
			"Content-Type": "application/json",
			...attribution(),
		},
		body: JSON.stringify(body),
	});
	if (!res.ok) throw await explain(res);

	const data = await res.json();
	return String(data?.choices?.[0]?.message?.content ?? "");
}

File: llm/providers/openrouter.py
"""OpenRouter: one key, many providers' models, billed to the user's account.

Every request OpenRouter gets from this app carries the app's attribution
headers, which is how its usage is credited to Clips Kitty in OpenRouter's app
rankings. They are the app's identity, not a user setting, and they are built
here and nowhere else: SPEC.extra_headers hands them to llm/providers/http.py,
which puts them on every request (chat, models, key check, transcription and
every retry). `video-gen` is the category in OpenRouter's Creative section
that fits; a bare "creative" is not a category and would be dropped.

The website (web/lib/openrouter.ts) sends its own set, with its own origin and
the title "Clips Kitty Web", from the visitor's browser.
"""

from llm.providers.base import ModelInfo, ProviderSpec

SITE = "https://openrouter.ai"
APP_URL = "https://colingpt9.github.io/clips-studio/"
APP_TITLE = "Clips Kitty"
APP_CATEGORY = "video-gen"


def attribution_headers() -> dict:
    return {
        "HTTP-Referer": APP_URL,
        "X-OpenRouter-Title": APP_TITLE,
        "X-OpenRouter-Categories": APP_CATEGORY,
    }


def _route_to_capable_providers(body: dict) -> dict:
    """With a response format set, only route to providers that honour it
    (OpenRouter's own advice for structured output); otherwise the request
    could land on one that ignores the format and returns prose."""
    if "response_format" in body:
        body = {**body, "provider": {**body.get("provider", {}), "require_parameters": True}}
    return body


def model_info(entry: dict) -> ModelInfo | None:
    """A model worth offering: text out, and able to answer in JSON."""
    model_id = str(entry.get("id") or "")
    arch = entry.get("architecture") or {}
    outputs = arch.get("output_modalities") or []
    if not model_id or (outputs and "text" not in outputs):
        return None
    params = set(entry.get("supported_parameters") or [])
    if not params & {"response_format", "structured_outputs"}:
        return None
    note = ""
    if model_id.endswith(":free"):
        note = "Free, with a small daily request limit; a long video can use it up."
    elif "contributor" in model_id:
        note = "Cheaper, but the provider may use what you send to improve its products."
    pricing = entry.get("pricing") or {}
    lines = _token_lines(pricing)
    return ModelInfo(
        id=model_id,
        name=str(entry.get("name") or model_id),
        context=int(entry.get("context_length") or 0),
        json_schema="structured_outputs" in params,
        tools="tools" in params,
        note=note,
        vendor=model_id.split("/", 1)[0] if "/" in model_id else "",
        price=_price(pricing),
        pricing=lines,
        free=_is_free(pricing),
        page_url=f"{SITE}/{model_id}",
    )


# The fields OpenRouter prices a text model by, as it names them, and how each
# is shown. Only those it actually lists for a model, above zero, are shown.
_TOKEN_FIELDS = (
    ("prompt", "Input", 1_000_000, "per 1M tokens"),
    ("completion", "Output", 1_000_000, "per 1M tokens"),
    ("input_cache_read", "Cached input", 1_000_000, "per 1M tokens"),
    ("input_cache_write", "Cache write", 1_000_000, "per 1M tokens"),
    ("internal_reasoning", "Reasoning", 1_000_000, "per 1M tokens"),
    ("request", "Per request", 1, "per request"),
    ("image", "Image input", 1, "per image"),
    ("web_search", "Web search", 1, "per search"),
)


def _amount(pricing: dict, field: str) -> float | None:
    """A listed price in USD, or None: absent, unreadable, or negative ("varies")."""
    try:
        value = float(pricing[field])
    except (KeyError, TypeError, ValueError):
        return None
    return value if value >= 0 else None


def _token_lines(pricing: dict) -> tuple[tuple[str, float, str, bool], ...]:
    lines = []
    for field, label, scale, unit in _TOKEN_FIELDS:
        value = _amount(pricing, field)
        if value:
            lines.append((label, round(value * scale, 6), unit, False))
    return tuple(lines)


def _is_free(pricing: dict) -> bool:
    """Listed at zero for input and output, today. OpenRouter can change that."""
    return _amount(pricing, "prompt") == 0 and _amount(pricing, "completion") in (0, None) \
        and not any(_amount(pricing, f) for f, *_ in _TOKEN_FIELDS)


# ---- voice (transcription) models ------------------------------------------------

# OpenRouter lists a speech model's audio price with no unit. Every per-second
# price in the catalogue sits below this (whisper-1 is 0.0001 a second, which
# is OpenAI's own $0.006 a minute); larger figures (0.1, 0.36) only make sense
# per hour, and nothing in the data says so. So a per-hour estimate is shown
# only below this line, labelled as an estimate, and never above it.
PER_SECOND_CEILING = 1e-3


def stt_model_info(entry: dict) -> ModelInfo | None:
    """A transcription model from OpenRouter's public speech catalogue."""
    model_id = str(entry.get("id") or "")
    outputs = (entry.get("architecture") or {}).get("output_modalities") or []
    if not model_id or "transcription" not in outputs:
        return None
    pricing = entry.get("pricing") or {}
    prompt, completion = _amount(pricing, "prompt"), _amount(pricing, "completion")
    lines: list[tuple[str, float, str, bool]] = []
    price = None
    if completion:
        # Token-billed, like a text model: shown exactly as listed.
        lines = list(_token_lines({"prompt": pricing.get("prompt"), "completion": pricing.get("completion")}))
        price = _price(pricing)
    elif prompt:
        lines.append(("OpenRouter rate", prompt, "unit not stated by OpenRouter", False))
        if prompt < PER_SECOND_CEILING:
            lines.append(("Per hour of audio", round(prompt * 3600, 4), "estimate", True))
    return ModelInfo(
        id=model_id,
        name=str(entry.get("name") or model_id),
        vendor=model_id.split("/", 1)[0] if "/" in model_id else "",
        price=price,
        pricing=tuple(lines),
        free=prompt == 0 and not completion,
        verified=model_id in SPEC.stt["models"],
        page_url=f"{SITE}/{model_id}",
    )


def _price(pricing: dict) -> tuple[float, float] | None:
    """OpenRouter lists USD per token as strings; shown per million tokens.
    A negative price means "varies" (a router model), which is not a price."""
    try:
        per_token = float(pricing["prompt"]), float(pricing["completion"])
    except (KeyError, TypeError, ValueError):
        return None
    if min(per_token) < 0:
        return None
    return round(per_token[0] * 1_000_000, 4), round(per_token[1] * 1_000_000, 4)


SPEC = ProviderSpec(
    id="openrouter",
    label="OpenRouter",
    adapter="chat_completions",
    base_url="https://openrouter.ai/api/v1",
    key_label="OpenRouter API key",
    key_url="https://openrouter.ai/keys",
    pricing_url="https://openrouter.ai/models",
    privacy="Transcripts and prompts (and audio, if it transcribes) go to OpenRouter, and on to the "
            "model's provider, with your key.",
    extra_headers=attribution_headers,
    body_extras=_route_to_capable_providers,
    # /models/user, as the website uses: only models this account can route
    # to, after its own privacy and provider settings.
    models_path="/models/user",
    model_filter=model_info,
    stt_filter=stt_model_info,
    key_check_path="/key",
    # The recommended cloud path: one key reaches many models and providers.
    tier=2,
    tagline="One API key for many AI models and providers.",
    # Whisper through OpenRouter, with word timings. Short parts: OpenRouter's
    # upstream providers stop after about a minute of processing a request.
    # "models" are the ones known to return the word timings captions need;
    # the full live list comes from "catalog", and any other model is checked
    # with a short test clip when the user picks it (transcription/cloud.py).
    stt={"format": "openrouter", "chunk_seconds": 180,
         "models": ["openai/whisper-large-v3-turbo", "openai/whisper-large-v3", "openai/whisper-1"],
         "catalog": "/models?output_modalities=transcription&limit=100"},
)
